### PR TITLE
util: reduce STW from ReadMemStats

### LINKED ISSUE #38271: manage runtime.ReadMemStats to avoid stop-the-world too much
## Enhancement

Now in the tidb repo, we have some places to use ``` runtime.ReadMemStats``` to get global memory useage. But using it will cause the stop-the-world. this will affect performance. we should manage its usage.

### PR DESCRIPTION
**Pull request by hawkingrei**
Signed-off-by: Weizhen Wang <wangweizhen@pingcap.com>

### What problem does this PR solve?

Issue Number: close #38271

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
Does any test need to be done on this commit?

**Comment by hawkingrei** [comment]
> Does any test need to be done on this commit?

OK, I will. it is in the benchmark.

**Comment by hawkingrei** [comment]
> Does any test need to be done on this commit?

I think it can be covered by using the existing test cases.

**Comment by hawkingrei** [comment]
image

it can improve a little bit of performance.

File: util/memory/memstats.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+func ReadMemStats() *runtime.MemStats {
+	s := stats.Load()
+	if s != nil && time.Since(s.ts) < 300*time.Millisecond {
+		return &s.m
+	}
+	var g globalMstats
+	g.ts = time.Now()
+	runtime.ReadMemStats(&g.m)
+	stats.Store(&g)
+	return &g.m
```
Comment (wshwsh12) [inline util/memory/memstats.go:35]: return ForceReadMemStats()?

File: executor/explain.go [hunk 2]
```diff
@@ -169,8 +169,7 @@ func (h *memoryDebugModeHandler) fetchCurrentMemoryUsage(gc bool) (heapInUse, tr
 	if gc {
 		runtime.GC()
 	}
-	instanceStats := &runtime.MemStats{}
-	runtime.ReadMemStats(instanceStats)
+	instanceStats := memory.ReadMemStats()
```
Comment (wshwsh12) [inline executor/explain.go:172]:
```suggestion
	instanceStats := memory.ForceReadMemStats()
```

File: util/memory/memstats.go (hunk 1 above)
Comment (hawkingrei) [reply util/memory/memstats.go]: Done.

File: util/gctuner/memory_limit_tuner.go [hunk 3]
```diff
@@ -45,8 +44,7 @@ func (t *memoryLimitTuner) tuning() {
 	if !t.isTuning.Load() {
 		return
 	}
-	r := &runtime.MemStats{}
-	runtime.ReadMemStats(r)
+	r := memory.ForceReadMemStats()
```
Comment (wshwsh12) [inline util/gctuner/memory_limit_tuner.go:47]: Can use readMemoryInuse() the following check and remove the variable

**Review by wshwsh12 — APPROVED** [review]
LGTM

**Review by xhebox — APPROVED** [review]

**Pull request merged**